video/decode/vd_lavc.c

```diff
@@ -583,8 +585,13 @@ static void init_avctx(struct dec_video *vd, const char *decoder,
     AVCodecContext *avctx = ctx->avctx;
     if (!ctx->avctx)
         goto error;
+    avctx->opaque = vd;
     avctx->codec_type = AVMEDIA_TYPE_VIDEO;
     avctx->codec_id = lavc_codec->id;
+    if ((lavc_codec->capabilities & AV_CODEC_CAP_DR1) && vd->vo->driver->get_buffer) {
+        avctx->get_buffer2 = get_buffer2_direct;
+        avctx->thread_safe_callbacks = vd->vo->driver->caps & VO_CAP_GET_BUFFER_THREADSAFE;
```

Comment:
More importantly, if `VO_CAP_GET_BUFFER_THREADSAFE` is not set, then you need to make sure to call `driver->get_buffer` from the VO thread. Also, I think the frame "pooling" should be done outside of the specific VOs, but in a higher level. Basically this code should directly give lavc a pre-mapped buffer from a pool of buffers (so it can return right away), and this pool should be generated using `driver->get_buffer` from the VO thread as needed. (Probably using a fixed size)